packages/components/src/TreeView/TreeViewItem/TreeViewItem.component.js

```diff
@@ -112,57 +125,202 @@ class TreeViewItem extends React.Component {
 	}
 
 	onKeyDown(event) {
-		if (event.keyCode === keycode.codes.enter) {
-			this.onSelect();
+		switch (event.keyCode) {
+			case keycode.codes.enter:
+			case keycode.codes.space:
+				this.onSelect(event);
+				break;
+			case keycode.codes.left:
+				if (this.hasChildren() && this.props.item.toggled) {
+					this.onToggle(event);
+				} else if (!this.hasChildren() || !this.props.item.toggled) {
+					focusOn(event, this.getParentItem());
+				}
+				break;
+			case keycode.codes.right:
+				if (this.hasChildren() && !this.props.item.toggled) {
+					this.onToggle(event);
+				} else if (this.hasChildren() && this.props.item.toggled) {
+					focusOn(event, this.getFirstChildItem());
+				}
+				break;
+			case keycode.codes.down:
+				focusOn(event, this.getNextItem());
+				break;
+			case keycode.codes.up:
+				focusOn(event, this.getPreviousItem());
+				break;
+			case keycode.codes.home:
+				focusOn(event, this.getFirstItem());
+				break;
+			case keycode.codes.end:
+				focusOn(event, this.getLastItem());
+				break;
+			case keycode.codes['numpad *']:
+				this.onToggleAllSiblings(event);
+				break;
+			default:
+				break;
 		}
 	}
 
-	onSelect() {
-		this.containerRef.focus();
-		this.props.onClick(this.props.item);
+	onToggleAllSiblings(event) {
+		event.stopPropagation();
+		return this.props.onToggleAllSiblings(this.props.itemSiblings);
+	}
+
+	onSelect(event) {
+		event.stopPropagation();
 		return this.props.onSelect(this.props.item);
 	}
 
-	renderTreeViewItem(child, i) {
+	onToggle(event) {
+		event.stopPropagation();
+		return this.props.onToggle(this.props.item);
+	}
+
+	getFirstItem() {
+		return this.containerRef.closest('ul[role="tree"]').querySelector('li[role="treeitem"]');
+	}
+
+	getLastItem() {
+		const nodes = this.containerRef
+			.closest('ul[role="tree"]')
+			.querySelectorAll('li[role="treeitem"]');
+		return nodes.item(nodes.length - 1);
+	}
+
+	getParentItem() {
+		return this.containerRef.parentElement.closest('li[role="treeitem"]');
+	}
+
+	getFirstChildItem() {
+		return this.containerRef.querySelector('li[role="treeitem"]');
+	}
+
+	getNextItem() {
+		let nextElement;
+		let currentFound;
+		let hasNext;
+
+		const nodes = this.containerRef
+			.closest('ul[role="tree"]')
+			.querySelectorAll('li[role="treeitem"]')
+			.values();
+
+		do {
+			const { value, done } = nodes.next();
+
+			if (currentFound) {
+				nextElement = value;
+				hasNext = false;
+			} else {
+				currentFound = value === this.containerRef;
+				hasNext = !done;
+			}
+		} while (hasNext);
+
+		return nextElement;
+	}
+
+	getPreviousItem() {
+		let previousElement;
```

Comment:
the difference between getPreviousItem and getNextItem has lots of repetitions. Could you refactor it please ?

Reply:
`width deep structure` example is broken
<img width="549" alt="screen shot 2018-09-05 at 13 55 22" src="https://user-images.githubusercontent.com/19857479/45091650-6b731800-b113-11e8-81fc-b01d274fdad5.png">

Reply:
`the difference between getPreviousItem and getNextItem has lots of repetitions. Could you refactor it please ?`
--> not really, there are changes that make it not really possible to factor it. The only thing that I can put in common is the selection of all items. I did that